### PR TITLE
KATI-50 [Server] [Re-factor] Create enum for mapping database column for User, Facility and Speciality

### PR DESCRIPTION
**Pull request by huuduy216**

### CODE REVIEW

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 1]
```diff
@@ -5,77 +5,130 @@
...
+			stmt = con.prepareStatement("INSERT INTO " + tableName + " (name) VALUES (" + entity.getName() + ")");
+			stmt.executeUpdate();
+			stmt = con.prepareStatement("SELECT FROM " + tableName + " WHERE name =" + entity.getName());
+			rs = stmt.executeQuery();
+			rs.next();
+			Entity result = getEntityFromResultSet(rs);
+			return result;
+		} catch (Exception e) {
+			rs.close();
+			stmt.close();
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: These 2 lines must be put into finally, not in catch

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 2]
```diff
@@ -5,77 +5,130 @@
...
 		ResultSet rs = null;
-		stmt = con.prepareStatement("INSERT INTO " + tableName + " (name) VALUES (" + entity.getName() + ")");
-		stmt.executeUpdate();
-		stmt = con.prepareStatement("SELECT FROM " + tableName + " WHERE name =" + entity.getName());
-		rs = stmt.executeQuery();
-		return rs;
+		try {
+			stmt = con.prepareStatement("INSERT INTO " + tableName + " (name) VALUES (" + entity.getName() + ")");
+			stmt.executeUpdate();
+			stmt = con.prepareStatement("SELECT FROM " + tableName + " WHERE name =" + entity.getName());
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Don't use the same variable for different purposes

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 3]
```diff
@@ -5,77 +5,130 @@
...
+			rs = stmt.executeQuery();
+			rs.next();
+			Entity result = getEntityFromResultSet(rs);
+			return result;
+		} catch (Exception e) {
+			rs.close();
+			stmt.close();
+		}
+		Entity result = getEntityFromResultSet(rs);
+		return result;
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Put these 2 lines into try block too

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+		PreparedStatement stmt = null;
+		UserEntity userEntity = super.insertUser(entity);
+		ResultSet rs = null;
+
+		if (!(entity instanceof DoctorUserEntity)) {
+			return null;
+		}
+		try {
+			con.setAutoCommit(false);
+			if (!(entity instanceof DoctorUserEntity)) {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Remove this redundant if

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			int userId = Integer.parseInt(userEntity.getId());
+			entity.setId(String.valueOf(userId));
+			stmt = getInsertStatementForUser((DoctorUserEntity) entity);
+			stmt.executeUpdate();
+
+			insertToDoctorReferenceTables((DoctorUserEntity) entity);
+			stmt.close();
+			con.commit();
+
+			stmt = con.prepareStatement("SELECT * FROM doctor WHERE " + DoctorColumn.USER_ID.getColumnName() + "= ? ");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Don't use the same variable for different purposes

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		} catch (SQLException e) {
+			if (con != null) {
+				System.err.print("Transaction is being rolled back");
+				con.rollback();
+			}
+		} finally {
+			con.setAutoCommit(true);
+			stmt.close();
+		}
+		return getEntityFromResultSet(rs);
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Should not return here, but instead throws exception in the catch above

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+		return createPreparedStatement(doctorValueMap, databaseName, databaseMethod);
+	}
+
+	protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQLException {
+		List<ReferenceTableRow> referenceRows = new ArrayList<ReferenceTableRow>();
+		for (FacilityEntity facility : entity.getFacility()) {
+			ReferenceTableRow referenceRow = new ReferenceTableRow();
+			referenceRow.setFirstRefKey("facility_id");
+			referenceRow.setFirstValue(facility.getId());
+			referenceRow.setSecondRefKey("doctor_id");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Should create FacilityDoctorColumn to hold "facility_id" and "doctor_id"

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		referenceRows.clear();
+		for (SpecialityEntity speciality : entity.getSpeciality()) {
+			ReferenceTableRow referenceRow = new ReferenceTableRow();
+			referenceRow.setFirstRefKey("speciality_id");
+			referenceRow.setFirstValue(speciality.getId());
+			referenceRow.setSecondRefKey("doctor_id");
+			referenceRow.setSecondValue(entity.getId());
+			referenceRows.add(referenceRow);
+		}
+		insertReferenceRowsToReferenceTable("Doctor_Specialties", referenceRows);
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Same here, should create DoctorSpecialityColumn to hold "speciality_id" and "doctor_id"

File: com.kms.cura.dal/src/com/kms/cura/dal/database/FacilityDatabaseHelper.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+		super();
+	}
+
+	@Override
+	protected Entity getEntityFromResultSet(ResultSet resultSet) throws SQLException, ClassNotFoundException {
+		List<OpeningHour> openingHours = new ArrayList<OpeningHour>();
+		ResultSet openingHoursRS = this.queryByReferenceID("opening_hours", "facility_id", resultSet.getInt(EntityColumn.ID.getColumnName()));
+		while (openingHoursRS.next()) {
+			openingHours.add(new OpeningHour(DayOfTheWeek.getDayOfTheWeek(openingHoursRS.getInt("week_day")),
+					openingHoursRS.getTime("time_open"), openingHoursRS.getTime("time_close")));
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/FacilityDatabaseHelper.java]: Put opening_hours, facility_id, week_day, time_open, time_close, into OpeningHourColumn

File: com.kms.cura.dal/src/com/kms/cura/dal/database/UserDatabaseHelper.java [hunk 10]
```diff
@@ -4,216 +4,58 @@
...
+				stmt.setString(1, entity.getEmail());
+				rs = stmt.executeQuery();
+				con.commit();
+				UserDatabaseHelper dbh = new UserDatabaseHelper();
+				rs.next();
+				return dbh.getEntityFromResultSet(rs);
+			} catch (SQLException e) {
+				if (con != null) {
+					System.err.print("Transaction is being rolled back");
+					con.rollback();
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/UserDatabaseHelper.java]: After rollback, throws exception to return error to the apis

**Comment by huuduy216** [comment]
All mentioned issues have been fixed, please verify them.

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+				System.err.print("Transaction is being rolled back");
+				con.rollback();
+				throw e;
+			}
+		} finally {
+			con.setAutoCommit(true);
+			stmt.close();
+			stmt2.close();
+		}
+		return getEntityFromResultSet(rs);
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Fix this be moving throw e in catch above

**Comment by huuduy216** [comment]
Fixed additional issue

File: com.kms.cura.dal/src/com/kms/cura/dal/database/FacilityDatabaseHelper.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	@Override
+	protected Entity getEntityFromResultSet(ResultSet resultSet) throws SQLException, ClassNotFoundException {
+		List<OpeningHour> openingHours = new ArrayList<OpeningHour>();
+		ResultSet openingHoursRS = null;
+		try {
+			openingHoursRS = this.queryByReferenceID("opening_hours", "facility_id",
+					resultSet.getInt(EntityColumn.ID.getColumnName()));
+			while (openingHoursRS.next()) {
+				openingHours.add(new OpeningHour(DayOfTheWeek.getDayOfTheWeek(openingHoursRS.getInt("week_day")),
+						openingHoursRS.getTime("time_open"), openingHoursRS.getTime("time_close")));
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/FacilityDatabaseHelper.java]: Should put opening_hours, facility_id, week_day, time_open, time_close into OpeningHourColumn

File: com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+			con.setAutoCommit(false);
+			int userId = Integer.parseInt(userEntity.getId());
+			entity.setId(String.valueOf(userId));
+			stmt = getInsertStatementForUser((PatientUserEntity) entity);
+			stmt.executeUpdate();
+			stmt.close();
+			con.commit();
+
+			stmt = con
+					.prepareStatement("SELECT * FROM Patient WHERE " + PatientColumn.USER_ID.getColumnName() + " = ? ");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java]: Put "Patient" into constant

File: com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+			stmt.setInt(1, userId);
+			rs = stmt.executeQuery();
+			rs.next();
+			return getEntityFromResultSet(rs);
+		} catch (SQLException e) {
+			if (con != null) {
+				System.err.print("Transaction is being rolled back");
+				con.rollback();
+				throw e;
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java]: move this outside of if

File: com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+			}
+		} finally {
+			con.setAutoCommit(true);
+			stmt.close();
+		}
+		return getEntityFromResultSet(rs);
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java]: remove this

File: com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	public PreparedStatement getInsertStatementForUser(PatientUserEntity entity) throws SQLException {
+		Map<String, Object> patientValueMap = getColumnValueMapForEntity(entity);
+		String databaseName = "Patient";
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java]: use constant for "Patient"

File: com.kms.cura.dal/src/com/kms/cura/dal/database/UserDatabaseHelper.java [hunk 17]
```diff
@@ -4,214 +4,72 @@
...
 			}
-			int userId = rs.getInt("id");
-			entity.setId(String.valueOf(userId));
-			stmt = getInsertStatementForUser(entity);
-			stmt.executeUpdate();
+			
+			sb.setLength(0);
+			sb.append("INSERT INTO ");
+			sb.append(UserColumn.TABLE_NAME.getColumnName());
+			sb.append(" (email,password) VALUES (?,?)");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/UserDatabaseHelper.java]: use enum for email and password

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+			default:
+				break;
+			}
+		}
+		return columnValueMap;
+	}
+
+	public PreparedStatement getInsertStatementForUser(DoctorUserEntity entity) throws SQLException {
+		Map<String, Object> doctorValueMap = getColumnValueMapForEntity(entity);
+		String databaseName = "Doctor";
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Use constant for "Doctor"

**Comment by huuduy216** [comment]
Fixed new issues

**Comment by huuduy216** [comment]
Change table name mapping structure

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 19]
```diff
@@ -5,77 +5,133 @@
...
-	public ResultSet queryByID(String tableName, int id) throws SQLException {
+	public Entity queryByID(String tableName, int id) throws SQLException, ClassNotFoundException {
+		PreparedStatement stmt = null;
 		ResultSet rs = null;
-		stmt = con.prepareStatement("SELECT * FROM " + tableName + " WHERE id = ?");
-		stmt.setInt(1, id);
-		rs = stmt.executeQuery();
-		return rs;
+		try {
+			stmt = con.prepareStatement("SELECT * FROM " + tableName + " WHERE id = ?");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Change "id" into UserColumn.ID

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 20]
```diff
@@ -5,77 +5,133 @@
...
-	public ResultSet queryByName(String tableName, String name) throws SQLException {
+	public Entity queryByName(String tableName, String name) throws SQLException, ClassNotFoundException {
+		PreparedStatement stmt = null;
 		ResultSet rs = null;
-		stmt = con.prepareStatement("SELECT * FROM " + tableName + " WHERE name = ?");
-		stmt.setString(1, name);
-		rs = stmt.executeQuery();
-		return rs;
+		try {
+			stmt = con.prepareStatement("SELECT * FROM " + tableName + " WHERE name = ?");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Change "name" into UserColumn.Name

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 21]
```diff
@@ -5,77 +5,133 @@
...
+		PreparedStatement stmt = null;
+		PreparedStatement stmt2 = null;
 		ResultSet rs = null;
-		stmt = con.prepareStatement("INSERT INTO " + tableName + " (name) VALUES (" + entity.getName() + ")");
-		stmt.executeUpdate();
-		stmt = con.prepareStatement("SELECT FROM " + tableName + " WHERE name =" + entity.getName());
-		rs = stmt.executeQuery();
-		return rs;
+		try {
+			stmt = con.prepareStatement("INSERT INTO " + tableName + " (name) VALUES (" + entity.getName() + ")");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Change "name" into UserColumn.Name

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java [hunk 22]
```diff
@@ -5,77 +5,133 @@
...
+			stmt.executeUpdate();
+			
+			stmt2 = con.prepareStatement("SELECT FROM " + tableName + " WHERE name =" + entity.getName());
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DatabaseHelper.java]: Change "name" into UserColumn.Name

File: com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+		Map<String, Object> patientValueMap = getColumnValueMapForEntity(entity);
+		String databaseName = PatientColumn.TABLE_NAME;
+		String databaseMethod = "INSERT INTO";
+		return createPreparedStatement(patientValueMap, databaseName, databaseMethod);
+	}
+
+	@Override
+	protected UserEntity getEntityFromResultSet(ResultSet resultSet) throws SQLException, ClassNotFoundException {
+		UserDatabaseHelper dbh = new UserDatabaseHelper();
+		UserEntity userEntity = (UserEntity) dbh.queryByID("Users",
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/PatientUserDatabaseHelper.java]: Change "Users" into constant

**Comment by huuduy216** [comment]
fixed

**Comment by huuduy216** [comment]
More refractory added

**Comment by hieumai** [comment]
OK from me. @vtnlinh95 and @toannguyen71195  please go ahead and verify this.

**Comment by vtnlinh95** [comment]
Verified ok

**Comment by toannguyen71195** [comment]
Verified

**Pull request merged**